File: 13-Algorithms_sorting/sorting/lib/bubble_sort.ex
defmodule BubbleSort do
  def sort([]), do: []
  def sort([x]), do: [x]
  def sort(lst) when not is_list(lst), do: nil

  def sort(lst) do
    case swap(lst) == lst do
      true -> swap(lst)
      false -> sort(swap(lst))
    end
  end

  defp swap([h, s | t]) when h > s, do: [s | swap([h | t])]
  defp swap([h, s | t]), do: [h | swap([s | t])]
  defp swap(lst), do: lst
end

# BubbleSort.sort([99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0])

File: 13-Algorithms_sorting/bubble_sort.py
def bubble_sort(lst):
    if len(lst) < 2:
        return lst

    for _i in range(len(lst) - 1):
        for j in range(len(lst) - 1):
            if lst[j] > lst[j + 1]:
                lst[j], lst[j + 1] = lst[j + 1], lst[j]
    return lst


numbers = [99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0]
print(bubble_sort(numbers))
